### PR TITLE
Adds gif support to reader detail content and reader comments

### LINKED ISSUE #9253: Animated GIF support in the reader detail screen
Let's add support for animated GIFs in the reader detail screen (minding the performance/memory impact of doing so!)

### LINKED ISSUE #9331: Animated gif support in reader comments
Let's add support for animated GIFs to reader comments (viewing only). 

We can probably leverage what will be updated in Reader Detail (#9253): `WPRichTextImage` to display either a static image or an animated gif.

Ref: #6062

### PR DESCRIPTION
**Pull request by bummytime**
This is an almost-complete PR that adds gif support to the reader detail content, reader detail full screen preview (`WPImageViewController`), and the reader comments:

gif_reader2

It's basically ready for you to start taking a look at @etoledom.

Instead of using `WPTableImageSource` to load the images, `WPRichTextImage` now uses our new `ImageLoader` for all images (animated or not).  `ImageLoader` was updated to to now allow a photon quality var to be set — which is `65` for within the reader detail (and defaults to `80` everywhere else like previously).

Trying to reduce memory is the name of the game here. I put in a few safeguards like:

* Selecting the `small` gif playback strategy
* `AnimatedImageCache` is now listening for memory warnings and dumping the cache if it occurs
* As mentioned ☝️, `ImageLoader` sets photon quality to 65
* Clearing data from `CachedAnimatedImageView` as soon as possible

Even with all that ☝️, memory usage **is** higher than our previous static, non-animated gif display (assuming you view a lot of reader posts containing gifs). The question is if this implementation is acceptable in terms of memory usage.

Fixes #9253 
Fixes #9331  

## WIP

1. Like the previous reader image loading implementation, there isn't an loading indicator for images. The image size is set AFTER the image is loaded. I am not sure if this is best approach now, and will most likely add the indicator back in. 😄 

2. There is a image sizing issue while scrolling through reader comments. I'll patch it asap.

## To Test:

**Note 1:** Try to test on device since that will give us an accurate gauge of mem usage.

**Note 2:** Be sure to also test a slow network, not just wifi.

**Note 3:** iPads!!!  (Ok, ok no more notes!)

1. Locate reader posts with gifs, heck, locate reader posts with **ALL** kinds of media. Make sure it all works. Hawk the memory usage while you are doing this.

2. Next, find some reader comments that contain gifs (ask me if you need help finding some 😄 ). Verify they display correctly.

### CODE REVIEW

**Comment by etoledom** [comment]
As a first glance, this look WAY TOO COOL! (sorry for the excitement 😆)

Seriously, I was sad because we decided to not autoplay gifs in the post detail, but event testing in the iPhone 5 it scrolls smooth (once the gifs are loaded).

Agreed that memory could be an issue. I'll continue testing with edge cases in the old phone.

Great stuff @bummytime !

**Comment by bummytime** [comment]
@etoledom I added the feature flag for gif support in reader detail. Let me know how your smoke tests go on older devices.

**Comment by etoledom** [comment]
Since @bummytime is AFK for some days, I took over this PR, and made a couple of extra changes.

The most important one is to use Photon to resize Gif images and load smaller version of them.
Sadly, this is not always the best, since the smaller the original gif image is, the less quality seem to be the final result. 

To solve that problem, the image will be resized just if its width is bigger than the device's width.

In this way, if there are big gif images, they will be resized to be smaller, and the lost of quality also help in reducing the memory usage and the final quality is not so noticeable.

I also noticed that the Photon `image quality` property doesn't affect gif images.

**Note:**
There are plans to replace the Reader Detail with a new Table View. That will help a lot in the memory usage of gif images, thanks to the lazy loading of the images. But even so, this PR will not be a lost work, since there is many gif performance tuning that all other sections that use gifs can benefit from.

@aerych Could you take a look please? I believe you have work in this section before :)
@frosty This might be of interest for you too, to implement gifs in the future new `ReaderDetail`.

---
**To test:**

- Go to the reader.
- Go back to the Reader Streams selector and choose `Search`.
- Search for `gif`.
- Open [Your WIP in Gifs Tag!](https://wordpress.com/read/blogs/53534285/posts/6148)
- Check that gifs are loading and animating in the body of the post.
---

- Go back to the search result and scroll down.
- Look for: [Brilliantly Creepy Gifs From Old Photographs by Kevin Weir](https://wordpress.com/read/blogs/127597222/posts/43271)
  - This post has a lot of very big gifs, and is the one we have been testing.
- Check that everything work just fine.
---

- If you can, please test in old devices (I tested in a `iPhone 5`)
- If posible too, test on an iPad Pro 12.9'', since the size of the gifs loaded will depend on screen size.
---

- Test disabling the feature flag and check that it behaves as it did before, loading static images.

**Comment by etoledom** [comment]
Thank you @aerych for the review!
I made the changed requested and it's ready for another look.

Since @aerych is going to vacations until Tuesday, @frosty could you please take a final look?

I believe that checking out that there are no layout problems due to the last change will be enough, since the rest was already reviewed 👍

File: WordPress/Classes/ViewRelated/Views/WPRichText/WPRichTextImage.swift [hunk 1]
```diff
@@ -56,18 +62,57 @@ open class WPRichTextImage: UIControl, WPRichTextMediaAttachment {
...
+            return CGSize(width: 1.0, height: 1.0)
         }
-        return CGSize(width: 1.0, height: 1.0)
+        return imageView.intrinsicContentSize
     }
 
     func contentRatio() -> CGFloat {
-        if let size = imageView.image?.size {
-            return size.width / size.height
+        guard imageView.intrinsicContentSize != .zero else {
```
[the remark this answers is not in the document]
Comment (aerych) [reply WordPress/Classes/ViewRelated/Views/WPRichText/WPRichTextImage.swift]: I think its something we (meaning probably me) just missed and forgot to clean up in a previous refactor. :)

**Comment by etoledom** [comment]
Thank you @aerych ! 
And sorry for making you work in your day off 😅

**Pull request merged**